src/Utils.php

```diff
@@ -700,30 +686,69 @@ function _wf_civicrm_api3_contribution_transact_spec(&$params) {
+      $params['invoice_id'] = md5(uniqid(rand(), TRUE));
+    }
+    if (isset($params['invoice_id']) && !isset($params['invoiceID'])) {
+      // This would be required prior to https://lab.civicrm.org/dev/financial/issues/77
       $params['invoiceID'] = $params['invoice_id'];
     }
+    elseif (!isset($params['invoice_id']) && isset($params['invoiceID'])) {
+      $params['invoice_id'] = $params['invoiceID'];
+    }
+
+    $order = civicrm_api3('Order', 'create', $params);
+    try {
+      // Use the Payment Processor to attempt to take the actual payment. You may
+      // pass in other params here, too.
+      $propertyBag = new \Civi\Payment\PropertyBag();
+      $propertyBag->mergeLegacyInputParams($params);
+      $propertyBag->setContributionID($order['id']);
+      // @fixme: Class \Civi\Payment\CRM_Utils_Money not found (fixed in 5.28 via https://github.com/civicrm/civicrm-core/pull/17505
+      //   But note we still return a localized format
+      // $propertyBag->setAmount($params['total_amount']);
+
+      $reflectionClass = new ReflectionClass($propertyBag);
+      $reflectionProperty = $reflectionClass->getProperty('props');
+      $reflectionProperty->setAccessible(TRUE);
+      $payParams = $reflectionProperty->getValue($propertyBag)['default'];
+
+      // propertyBag has 'contributionID', we need 'contribution_id'.
+      $payParams['contribution_id'] = $propertyBag->getContributionID();
+      $payParams['amount'] = $params['total_amount'];
+      $payResult = reset(civicrm_api3('PaymentProcessor', 'pay', $payParams)['values']);
+
+      // webform_civicrm sends out receipts using Contribution.send_confirmation API if the contribution page is has is_email_receipt = TRUE.
+      // We allow this to be overridden here but default to FALSE.
+      $params['is_email_receipt'] = $params['is_email_receipt'] ?? FALSE;
+
+      // Assuming the payment was taken, record it which will mark the Contribution
+      // as Completed and update related entities.
+      civicrm_api3('Payment', 'create', [
+        'contribution_id' => $order['id'],
+        'total_amount' => $payParams['amount'],
+        'payment_instrument_id' => $order['values'][$order['id']]['payment_instrument_id'],
+        // If there is a processor, provide it:
+        'payment_processor_id' => $payParams['payment_processor_id'],
+        'is_send_contribution_notification' => $params['is_email_receipt'],
+        'trxn_id' => $payResult['trxn_id'] ?? NULL,
+      ]);
+    }
+    catch (Exception $e) {
+      return ['error_message' => $e->getMessage()];
+    }
 
-    // Some payment processors expect a unique invoice_id - generate one if not supplied
-    $params['invoice_id'] = \CRM_Utils_Array::value('invoice_id', $params, md5(uniqid(rand(), TRUE)));
-
-    $paymentProcessor = \CRM_Financial_BAO_PaymentProcessor::getPayment($params['payment_processor'], $params['payment_processor_mode']);
-    $paymentProcessor['object']->doPayment($params);
-
```

Comment:
No more `doPayment`?

Reply:
No, it's called via `PaymentProcessor.Pay` API.